subsys/net/lib/mqtt_helper/mqtt_helper.c

```diff
@@ -383,7 +384,24 @@ static int broker_init(struct sockaddr_storage *broker,
 	addr = result;
 
 	while (addr != NULL) {
-		if (addr->ai_addrlen == sizeof(struct sockaddr_in)) {
+		/* Check for IPv6 first and use if available. */
+		if (addr->ai_addrlen == sizeof(struct sockaddr_in6)) {
+			struct sockaddr_in6 *broker6 = ((struct sockaddr_in6 *)broker);
+			char ipv6_addr[INET6_ADDRSTRLEN];
+
+			memcpy(broker6->sin6_addr.s6_addr,
+			       ((struct sockaddr_in6 *)addr->ai_addr)
+			       ->sin6_addr.s6_addr,
+			       sizeof(struct in6_addr));
+			broker6->sin6_family = AF_INET6;
+			broker6->sin6_port = htons(CONFIG_MQTT_HELPER_PORT);
+
+			inet_ntop(AF_INET6, &broker6->sin6_addr.s6_addr,
+				  ipv6_addr, sizeof(ipv6_addr));
+			LOG_DBG("IPv6 Address found %s", ipv6_addr);
+			break;
+		} else if (addr->ai_addrlen == sizeof(struct sockaddr_in)) {
```

Comment:
I will suggest adding a check if `CONFIG_NET_IPV4` is enabled, because when asking for `AF_UNSPEC` you may get both IPv4 and IPv6 address, and if `CONFIG_NET_IPV4` is not enabled the connection will fail later. I think this is a bug in Zephyr `getaddrinfo()`, and will make a bug fix PR for this, but it should also be checked here.

```suggestion
		if ((addr->ai_family == AF_INET) && IS_ENABLED(CONFIG_NET_IPV4)) {
```